File: Penna_Simulation/src/Penna/animal.cpp
/* ===================================================================
 * Programming Techniques for Scientific Simulations I  
 * package  :   Penna
 * function :   represent an animal for penna simulation
 * created  :   18/12/2023
 * author   :   DaeBlicki
 * 
 * updated  :   xx/xx/xx
 * VERSION 1.00 (C) Copyright Free to use
 * ===================================================================*/

#include "../../include/Penna/animal.hpp"
#include <stdexcept>

namespace Penna{

/* Definitions of static members */
unsigned int Animal::deletion_threshhold_;   /* maximal number of deletion of animal */
age_t Animal::reproduction_age_;             /* minimal age for reproduction of animal */
double Animal::propability_to_get_pregnant_; /* percentage to get pregnant */

Animal::Animal() : gen_(Penna::Genome())
{
    age_ = 0;
    pregnant_ = false;
}

Animal::~Animal()
{
}

Animal::Animal( const Animal& parent) : gen_(parent.gen_)
{
    age_ = parent.age_;
    pregnant_ = parent.pregnant_;
}

Animal::Animal( const Genome& gen, const age_t age, const bool pregnancy) : gen_(gen)
{
    age_ = age;
    pregnant_ = pregnancy;
}

bool Animal::operator<( const Animal& compared )
{
    return (age_ < compared.age());
}

bool Animal::operator==( const Animal& compared )
{
    return (age_ == compared.age());
}

bool Animal::operator>( const Animal& compared )
{
    return (age_ < compared.age());
}

age_t Animal::age() const
{
    return age_;
}

bool Animal::isPregnant() const
{
    return pregnant_;
}

Genome Animal::genome() const
{
    return gen_;
}

Animal Animal::giveBirth()
{
    if(pregnant_ != true){
        throw std::runtime_error("Animal must be pregnant to giving birth");
    }
    pregnant_ = false;
    age_t baby_age = 0;
    Genome baby_genome = Genome(gen_);
    baby_genome.mutate();
    return Animal(baby_genome, baby_age, false);
}

void Animal::reproduction()
{
    // try to reproduce, only if is old enough and NOT pregnant ;)
    if(age_ >= reproduction_age_ && !pregnant_){
        double pregnancy_luck = pt1::Random_Number_Generator::generate() % 100;
        pregnancy_luck = pregnancy_luck/ 100;
        if(propability_to_get_pregnant_ > pregnancy_luck){
            pregnant_ = true;
        }
    }
}

void Animal::growth()
{
    if(age_ == maximal_age_){
        throw std::runtime_error("Animal should never exceed the maximal age! \n");
    }
    age_++;
}


bool Animal::survive( const double verhulst)
{
    /* death by mutation */
    if(gen_.countBad(age_) > deletion_threshhold_){
        return false;
    }
    /* death by verhulst */
    double chance_to_die = pt1::Random_Number_Generator::generate() % 100;
    chance_to_die = chance_to_die/100;
    if(verhulst < chance_to_die){
        return false;
    }
    /* if maxiaml age arrived, is it time to go :(.. */
    if(age_ == maximal_age_){
        return false;
    }
    return true;
}

void Animal::init( const unsigned int deletion_treshhold, const age_t reproduction_age, const double probability_to_get_pregnant)
{
    deletion_threshhold_ = deletion_treshhold;
    if(reproduction_age >= maximal_age_){
       throw std::runtime_error("Reproduction Age exceed Maximal Age, which is absolute nonesense! \n");
    }
    reproduction_age_ = reproduction_age;
    if(probability_to_get_pregnant < 0 || probability_to_get_pregnant > 1){
       throw std::runtime_error("Probability to get pregnant chance is not in range [0,1]! \n");
    }
    propability_to_get_pregnant_ = probability_to_get_pregnant;
}

std::tuple<unsigned int, age_t, double> Animal::getInitialization()
{
    return std::tuple<unsigned, age_t, double>(deletion_threshhold_, reproduction_age_, propability_to_get_pregnant_);
}


age_t Animal::maximalAge()
{
    return Animal::maximal_age_;
}


} // end namespace penna

File: pyPenna/src/animal.py
# ============================================================
# Programming Techniques for Scientific Simulations I
# package   : pyPenna
# function  : represent animal class for pyPenna
# created   : 27/01/24
# author    : DaeBlicki
#
# VERSION 1.00 (C) Copyright Free
# ============================================================

import numpy as numpy
import random
from .genome import Genome

class Animal():
    """Represent an animal in pyPenna."""
    mutation_tolerance = 5          # maximal number of mutations [1, 63]
    reproduction_age = 12           # minimal age to reproduce [0, 63]
    chance_to_get_pregnant = 0.5    # chance to get pregnant [0, 1]
    maximal_age = 63                # maximal age for an animal

    def __init__(self, genome=None, age=0, pregnant=False):
        """Initialize animal object."""
        self.gen = Genome(genome)
        self.age = age
        self.pregnant = pregnant

    def init(self, mutation_tolerance, reproduction_age, 
             chance_to_get_pregnant):
        """Initialize member parameter shared by all animals.

        :param mutaton_tolerance     : integer in range [0, 63]
        :param reproduction_age      : integer in range [0, 63]
        :param chance_to_get_pregant : double in range [0, 1]
        """
        self.__class__.mutation_tolerance = mutation_tolerance
        self.__class__.reproduction_age = reproduction_age
        self.__class__.chance_to_get_pregnant = chance_to_get_pregnant

    def giveBirth(self):
        """Giving birth and return an animal baby with mutated genes
        
        :rtype : Animal
        :return: Newborn animal with mutated genes from mother
        """
        # animal can only give birth if it pregnant
        if self.pregnant:
            self.pregnant = False
            baby_genome = self.gen
            baby_genome.mutate
            return Animal(baby_genome)
        
    def reproduction(self):
        """Animal is trying to impregnate himself assexually

        :rtype : void
        :return: self has chance to get pregnant
        """
        # only check unpregnant and adult animals
        if(not self.pregnant and self.age >= self.__class__.reproduction_age):
            chance = random.random()
            if chance > self.chance_to_get_pregnant:
                self.pregnant = True
    
    def growth(self):
        """Animal is getting older
        
        :rtype : void
        :return: self is one year older
        """
        self.age += 1

    def survive(self, verhulst):
        """Animal is trying to survive the environment and mutations

        :param verhulst: environmental factor, double in range (0, 1]
        :rtype : bool
        :return: True, if the animal survived
        """
        # check biological factors
        if self.gen.countBad() > self.__class__.mutation_tolerance:
            return False
        # check age factor
        if self.age > self.__class__.maximal_age:
            return False
        # check environmental factor
        chance_to_die = random.random()
        if verhulst < chance_to_die:
            return False
        # Animal survived!
        return True


